[DTensor] support split op on Partial placement

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #158112
* #158051
* __->__ #157991

**Summary**
To enable use case where the input DTensor to `split` op has `Partial()` placement,
this PR treats `Partial()` in the same way with `Replicate()`. That means, `split` op
only unshards the `Shard(dim=x)` if `x == split_dim` and keep other placement
untouched.

**Test**
Added a new test because `test_dtensor_ops` doesn't test `Partial()` placement.
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @tianyu-l

XilunWu committed:
[DTensor] support split op on Partial placement

[ghstack-poisoned]

ezyang commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -708,6 +709,46 @@ def test_slice(self):
+        except AssertionError:
+            print(
+                f"failing case reduce_op={reduce_op} split_size={split_size} split_dim={split_dim}"
+            )
```
I don't understand, what's the point of this test case then?

ezyang replied:
The other contents seem fine though

ezyang commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -708,6 +709,46 @@ def test_slice(self):
+
+        partial_tensor = torch.randn(8, 8, device=self.device_type)
+        replicate_tensor = partial_tensor.detach().clone()
+        replicate_tensor = funcol.all_reduce(
```
A more generic testing thought: it seems to me that this sort of "do the operation with dtensor but also do it with plain tensor" feels like it would be pretty common, can we not automate it?  For example, we don't have to manually setup the replicate_tensor and collective: instead, we can simply do partial_dt.full_tensor().split(split_size, dim=split_dim). We're simply testing that we check this diagram commutes:

```
             split()
  partial_dt ---------> B
      |                  |
      |                  | full_tensor()
full_tensor()       result
      |                  |
      v                  v
      C ---------> result
             split()
```

XilunWu replied:
Good point, will add a base PR to introduce such test util. IIRC we have similar logics in each separate test file but haven't merged them into one shared util.

XilunWu replied:
implemented @ezyang's suggestion in https://github.com/pytorch/pytorch/pull/158112

XilunWu committed:
Update on "[DTensor] support split op on Partial placement"


**Summary**
To enable use case where the input DTensor to `split` op has `Partial()` placement,
this PR treats `Partial()` in the same way with `Replicate()`. That means, `split` op
only unshards the `Shard(dim=x)` if `x == split_dim` and keep other placement
untouched.

**Test**
Added a new test because `test_dtensor_ops` doesn't test `Partial()` placement.
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

zpcore commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -708,6 +710,41 @@ def test_slice(self):
+
+    def _test_split_on_partial(self, reduce_op: str, split_size: int, split_dim: int):
+        torch.manual_seed(self.rank)
+        mesh = init_device_mesh(self.device_type, (self.world_size,))
```
Can we try on 2d mesh just to be safe with the allreduce?

zpcore reviewed (approved):
LGTM!

XilunWu committed:
Update on "[DTensor] support split op on Partial placement"


**Summary**
To enable use case where the input DTensor to `split` op has `Partial()` placement,
this PR treats `Partial()` in the same way with `Replicate()`. That means, `split` op
only unshards the `Shard(dim=x)` if `x == split_dim` and keep other placement
untouched.

**Test**
Added a new test because `test_dtensor_ops` doesn't test `Partial()` placement.
`pytest test/distributed/tensor/test_tensor_ops.py -s -k test_split_on_partial`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

XilunWu commented:
@pytorchbot merge
